### PR TITLE
fix vulkan texture rendering

### PR DESCRIPTION
**Pull request by phr00t**
# PR Details

Allows textures to render in Vulkan

## Related Issue

https://github.com/xenko3d/xenko/issues/415

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My change requires a change to the documentation.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by xen2** [comment]
Thanks! I'll test and merge once back from GDC in a few days.

**Comment by xen2** [comment]
Just checking, is this ready for review? (I remember seeing on Discord there was some cases still not covered, but not sure if this was before or after this PR)

**Comment by phr00t** [comment]
I've been keeping this pull request up to date with my fixes. The only issue I'm aware of is point lights don't provide shadows in Vulkan. I've had some rare startup crashes, but I haven't noticed them yet with the latest commit (specific bool created yadda yadda). However, I haven't been stress testing it. I'd say this is ready for review.

File: sources/engine/Xenko.Graphics/Vulkan/CommandList.Vulkan.cs [hunk 1]
```diff
@@ -24,6 +24,7 @@ public partial class CommandList
         private int framebufferAttachmentCount;
         private bool framebufferDirty = true;
         private Framebuffer activeFramebuffer;
+        private Texture dummyTexture;
```
Comment (xen2) [inline sources/engine/Xenko.Graphics/Vulkan/CommandList.Vulkan.cs]: Shouldn't this be part of `GraphicsDevice.Vulkan.cs` instead?

File: sources/engine/Xenko.Graphics/Vulkan/CommandList.Vulkan.cs [hunk 2]
```diff
@@ -42,6 +43,7 @@ public static CommandList New(GraphicsDevice device)
         private CommandList(GraphicsDevice device) : base(device)
         {
             Recreate();
+            dummyTexture = Texture.New2D(GraphicsDevice, 1, 1, PixelFormat.R8G8B8A8_UNorm_SRgb, TextureFlags.ShaderResource);
```
Comment (xen2) [inline sources/engine/Xenko.Graphics/Vulkan/CommandList.Vulkan.cs]: Shouldn't this be part of `GraphicsDevice.Vulkan.cs` instead?

Comment (xen2) [reply sources/engine/Xenko.Graphics/Vulkan/CommandList.Vulkan.cs]: Also, need to add a line to destroy it when `GraphicsDevice` is destroyed.

File: sources/engine/Xenko.Graphics/Vulkan/CommandList.Vulkan.cs (hunk 1 above)
Comment (phr00t) [reply sources/engine/Xenko.Graphics/Vulkan/CommandList.Vulkan.cs]: Yes, it should. Working on another issue with Vulkan (https://github.com/phr00t/xenko/issues/49), so I'm not able to get to this right at the moment.

File: sources/engine/Xenko.Graphics/Vulkan/CommandList.Vulkan.cs (hunk 2 above)
Comment (phr00t) [reply sources/engine/Xenko.Graphics/Vulkan/CommandList.Vulkan.cs]: Yes, it should. Working on another issue with Vulkan (https://github.com/phr00t/xenko/issues/49), so I'm not able to get to this right at the moment.

File: sources/engine/Xenko.Graphics/Vulkan/CommandList.Vulkan.cs (hunk 1 above)
Comment (phr00t) [reply sources/engine/Xenko.Graphics/Vulkan/CommandList.Vulkan.cs]: Did it when I was trying to fix that other issue. Help there would be greatly appreciated.

File: sources/engine/Xenko.Graphics/Vulkan/CommandList.Vulkan.cs (hunk 2 above)
Comment (phr00t) [reply sources/engine/Xenko.Graphics/Vulkan/CommandList.Vulkan.cs]: This should be done.

File: sources/engine/Xenko.Graphics/Vulkan/GraphicsDevice.Vulkan.cs [hunk 3]
```diff
@@ -471,6 +478,7 @@ private unsafe void ReleaseDevice()
             // Release fenced resources
             nativeResourceCollector.Dispose();
             DescriptorPools.Dispose();
+            if( dummyTexture != null ) dummyTexture.Dispose();
```
Comment (xen2) [inline sources/engine/Xenko.Graphics/Vulkan/GraphicsDevice.Vulkan.cs]: Space around if.
Please split it into two lines.

File: sources/engine/Xenko.Graphics/Vulkan/PipelineState.Vulkan.cs [hunk 4]
```diff
@@ -9,6 +9,7 @@
 using Xenko.Core.Collections;
 using Xenko.Core.Serialization;
 using Xenko.Shaders;
+using System.Threading;
```
Comment (xen2) [inline sources/engine/Xenko.Graphics/Vulkan/PipelineState.Vulkan.cs]: Usually `using` are ordered (System then custom such as Xenko, alphabetically is a bonus then)

File: sources/engine/Xenko.Graphics/Vulkan/CommandList.Vulkan.cs [hunk 5]
```diff
@@ -424,6 +422,9 @@ public unsafe void ResourceBarrierTransition(GraphicsResource resource, Graphics
                     texture = texture.ParentTexture;
 
                 // TODO VULKAN: Check for change
+                
+                // wait, are we ready to use this texture yet?
+                if (texture.ReadyForSampling == false) texture = GraphicsDevice.dummyTexture;
```
Comment (xen2) [inline sources/engine/Xenko.Graphics/Vulkan/CommandList.Vulkan.cs]: Please split into two lines.

File: sources/engine/Xenko.Graphics/Vulkan/GraphicsDevice.Vulkan.cs [hunk 6]
```diff
@@ -85,6 +85,27 @@ public BufferInfo(long fenceValue, SharpVulkan.Buffer buffer, DeviceMemory memor
...
+        /// <summary>
+        /// A minimal 1x1 texture used while bigger textures are loading.
+        /// </summary>
+        public Texture dummyTexture
+        {
+            get
+            {
+                while( _dummyTexture.NativeImageView == ImageView.Null ) 
+                {
+                    Thread.Sleep(1); // wait until this dummy texture is ready to be used
```
Comment (xen2) [inline sources/engine/Xenko.Graphics/Vulkan/GraphicsDevice.Vulkan.cs]: Why is this necessary?
In which case would it be accessed before being ready?
Having a while/sleep loop looks like a recipe for disaster (deadlock, etc.)

`dummyTexture` should be a simple private field and have deterministic lifetime (initialization/usage/destruction).

Comment (phr00t) [reply sources/engine/Xenko.Graphics/Vulkan/GraphicsDevice.Vulkan.cs]: It doesn't seem to be necessary on NVIDIA hardware, but it appears to be on AMD. I suspect even dummyTexture tries to get used before it is fully initialized on AMD hardware, leading to a memory access violation. This sleep holds off returning it until dummyTexture is ready, so a memory access violation should never happen on it. If you can think of a better way of handling this, please let me know... but I'm going with what works. Getting Vulkan to work on different hardware and driver implementations is dirty business, apparently.

Comment (xen2) [reply sources/engine/Xenko.Graphics/Vulkan/GraphicsDevice.Vulkan.cs]: Not sure how this has anything to do with GPU.

The initialization is done on CPU side with code `dummyTexture = Texture.New2D(this, 1, 1, PixelFormat.R8G8B8A8_UNorm_SRgb, TextureFlags.ShaderResource);`

If nothing is using it before that, this while should never be used.
If it us used before that, it means the initialization should be done earlier.

Comment (phr00t) [reply sources/engine/Xenko.Graphics/Vulkan/GraphicsDevice.Vulkan.cs]: The crash isn't from a null dummyTexture, but a memory access violation likely from dummyTexxture.NativeImageView == null. Other textures were not having a NativeImageView set soon enough before sampling (which is why we have a dummyTexture in the first place) and it appears on AMD that dummyTexture's NativeImageView was null for too long after its CPU initialization. Hard to verify this, though, my main development machine is Intel / NVIDIA. After making this change, things worked on different AMD test machines -- so this is all the theory on why this fix worked.

Comment (xen2) [reply sources/engine/Xenko.Graphics/Vulkan/GraphicsDevice.Vulkan.cs]: `Texture.New2D` should call `InitializeFromImpl` which should set the `NativeImageView`, so `NativeImageView` should be set as soon as texture is created.

I think this specific problem (memory access violation) is very likely due to something else (i.e. CPU/GPU sync or async resource allocation too late/freeing too early) and should be fixed very differently.

Comment (phr00t) [reply sources/engine/Xenko.Graphics/Vulkan/GraphicsDevice.Vulkan.cs]: If NativeImageView is set as soon as the texture is created, then this Thread.Sleep should never be called, and behavior wouldn't have changed due to this, and AMD hardware would still crash.

The only other change that could plausibly explain working AMD hardware is checking if VK_NV_external_memory_capabilities exists before requesting it, which is an OpenVR-required Vulkan extension. It should exist in the latest AMD drivers (although it didn't in old AMD drivers) & add it just as if it was always being added in older (and still crashing) Xenko builds...

Comment (phr00t) [reply sources/engine/Xenko.Graphics/Vulkan/GraphicsDevice.Vulkan.cs]: Hold up.... VK_NV_external_memory_capabilities

I'm guessing "NV" in that extension is referring to NVIDIA. Which means AMD wouldn't have it...

Comment (xen2) [reply sources/engine/Xenko.Graphics/Vulkan/GraphicsDevice.Vulkan.cs]: Yes, `VK_NV_*` extensions are NVIDIA only

**Comment by xen2** [comment]
Closing, will need to break this down into smaller PR.
